### DISCUSSION TITLE
Binary data / change whole response

### DISCUSSION THREAD

**Discussion by benperiton** [usebruno/bruno#449] [Q&A]
Hey!

Just wondering if it is possible to change the entire result from a request?

For instance, if I call https://graph.microsoft.com/v1.0/me/photo/$value I get a binary response, which doesn't show the actual image in the Response tab, but the raw data. What I would really like to do is pass it through btoa() and return a base64 representation of that image.

So I guess my question is; Can Post Response scripting return a new value to be used? Or is it purely for setting variables etc?

**Comment by helloanoop** [comment]
@benperiton Yes, Currently there is no [api support](https://docs.usebruno.com/scripting/javascript-reference.html#response) for modifying the request itself

>  I get a binary response, which doesn't show the actual image in the Response tab, but the raw data. 

Is it possible to share a screenshot of how this works in other clients, We can try to solve this problem in the app.

**Reply by benperiton** [reply]
I know insomnia also returns raw data, I was hoping I could use the post response script to return different data - ie I could get a base64 representation of the image.

Doing it in the app could potentially lead to vulnerability's I guess.

It's a niche use case really, but changing the response could be pretty cool

**Reply by helloanoop** [reply]
> but changing the response could be pretty cool

100% ! I think we should keep Bruno to be extensible. And enable devs to hack to the edges.

> Doing it in the app could potentially lead to vulnerability's I guess.

We can disable this behaviour by default, and optionally allow the user to override the check via a toggle in `bruno.json` Ex: https://github.com/usebruno/bruno/discussions/385#discussioncomment-7196674
```json
{
  "version": "1",
  "name": "bruno-testbench",
  "type": "collection",
  "scripts": {
    "filesystemAccess": {
      "allow": true
    },
    "modifyResponse": {
      "allow": true
    }
  }
}
```

This is in-fact a fantastic idea. This would allow people to write custom visualisers. 
Could be a quick way for devs to build mini dashboards

### Wild ideas (A bit too much, but still)
- I'd like to build a personal dashboard where a stuff runs every 5 mins and displays me if there were
  - any tweets around bruno that are going viral
  - current star count of bruno
  - show the most upvoted issues inside bruno
  - search the web and show any articles or videos that were created recently

Ideally, if A dev has to do it, we have to spin up a whole react and node stack and what not. It's a bit of a pain to sit and update stuff.

Anyways, I strongly feel some awesome possibilities could emerge from this.

Many Devs (myself included) don't like GUI tools for building interfaces via drag and drop. We'd rather sit and code the whole thing up. Bruno might offer a middle ground here.

**Reply by benperiton** [reply]
That would be awesome! A res.setResponse(res) {...} would be cool - esp if the response tab had a "save as " option, then you could still do raw and save as image or pdf or whatever, but without having to display it and worry about it in the app. 

Love the idea of a quick dashboard, commited with the code, run it and get some info back from the project

**Reply by helloanoop** [reply]
@benperiton I think you can modify response like this as seen here https://github.com/usebruno/bruno/issues/435#issuecomment-1754594172

**Reply by benperiton** [reply]
@helloanoop It doesn't seem that it is possible to overwrite res.body which would be my preferred way - changing the header doesn't really help as it is already coming back as image/jpeg - so changing that to text or whatever still leaves it not in decent format.

(Incidentally, I don't seem to be able to require libraries?)

```
const { btoa } = require("btoa");

let test = btoa(res.body);

console.log(test);
```

results in:

![image](https://github.com/usebruno/bruno/assets/104396/0c388ebf-985f-4071-b80e-49c9df3307a9)

**Reply by helloanoop** [reply]
I think you are importing it correctly. It should be 
```
const btoa = require("btoa")
```

**Reply by benperiton** [reply]
Ah yes, I was just going by the docs - should have tried that first! :)

But yes, if I could overwrite res.body that would be perfect
